### PR TITLE
Add cluster logs tests

### LINKED ISSUE #1940: Design and develop a test that analyzes the logs and indicates if there is any unexpected behavior
In this issue, we are looking to develop a test that, having as input the cluster log, is able to determine when there is an unexpected behavior. This could be, for example, two or more simultaneous integrity sync tasks, as we can see in the following log:

```
2021/09/23 14:54:56 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'str_upd'
2021/09/23 14:54:57 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'syn_a_w_m'
2021/09/23 14:54:58 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Agent-info sync] Starting
2021/09/23 14:54:58 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Agent-info sync] All chunks updated in wazuh-db in 0.073404s.
2021/09/23 14:55:00 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'syn_i_w_m_p'
2021/09/23 14:55:00 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Agent-info sync] Finished in 2.476s (23 chunks updated).
2021/09/23 14:55:03 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Integrity sync] Finished in 179.970s.      <-------
2021/09/23 14:55:07 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Integrity sync] Finished in 184.673s.      <-------
2021/09/23 14:55:07 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'syn_a_w_m_p'
2021/09/23 14:55:09 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'new_str'
2021/09/23 14:55:11 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'str_upd'
2021/09/23 14:55:11 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'syn_i_w_m_p'
2021/09/23 14:55:12 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'syn_a_w_m'
2021/09/23 14:55:12 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'syn_i_w_m'
2021/09/23 14:55:13 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Agent-info sync] Starting
2021/09/23 14:55:13 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Agent-info sync] All chunks updated in wazuh-db in 0.075327s.
2021/09/23 14:55:13 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Integrity check] Waiting to receive zip file from worker.
2021/09/23 14:55:15 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'new_file'
2021/09/23 14:55:17 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Agent-info sync] Finished in 4.041s (23 chunks updated).
2021/09/23 14:55:28 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Integrity sync] Finished in 205.112s.      <-------
2021/09/23 14:55:36 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B55_manager_19] [Main] Command received: b'file_upd'
```

### Tasks

- [x] Investigate several logs of different builds.
- [x] Establish all the cases in which the test should fail.
- [x] Develop the test.
- [x] Check that it works correctly with several logs, both ,legitimate and modified logs.

### PR DESCRIPTION
**Pull request by Selutario**
|Related issue|
|---|
| Closes #1940 |

## Description

This PR adds four new tests to analyze the cluster logs. Those tests are:
- **Test_cluster_connection:** It checks if all the nodes connect correctly to the master and if any worker disconnects at some point.
- **Test_cluster_error_logs:** Checks if the word `error` (case insensitive) appears in the cluster.log of any node. All errors are later attached in the `report.html`.
- **Test_cluster_logs_order:** Check in each worker that the logs for each task appear in the correct order. If any log is skipped or duplicated, the test fails. Useful for finding task overlapping or other similar problems. The logs found in case of error are attached in the report.html together with the expected logs, for each failed node.
- **Test_cluster_sync:** Checks if Integrity sync takes place multiple times in a row on a worker with the same number of files. This could mean that synchronization is not taking place, thus the test will fail.

## Configuration options

- `--artifacts_path`: Path where cluster logs can be found inside each node folder. It should follow the structure below:
    ```.
    ├── master
    │   └── logs
    │       └── cluster.log
    ├── worker_x
    │   └── logs
    │       └── cluster.log
    └── ...
    ```
- `--html=report.html`: Create a html report with the test results.
- `--self-contained-html`: Store all the necessary data for the report inside the html file.

## Logs example

A `report.html` can be created including the results of all these tests, together with the performance test developed in #2130:
image

The error logs of the `test_cluster_error_logs` test are attached in the details section:
image

The same happens with the incorrect order logs of the `test_cluster_logs_order` test:
image

### CODE REVIEW

File: deps/wazuh_testing/wazuh_testing/tools/utils.py [hunk 1]
```diff
@@ -12,6 +12,69 @@
...
+    def get_root(self):
+        """Get root node of current Node instance (the first node which has no parent).
+
+        Returns:
+            Node: root node of the current Node instance.
+        """
+        if self.parent_node:
+            return self.parent_node.get_root()
+        else:
+            return self
```
Comment (AdriiiPRodri) [inline deps/wazuh_testing/wazuh_testing/tools/utils.py:75]:
```suggestion
        return self
```

File: tests/performance/test_cluster/test_cluster_logs/test_cluster_connection/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+### Tests information
+
+| Number of tests | Time spent |
+|:--:|:--:|
+| 1 | 0.1s |
+
+## Expected behavior
+
+- Fail if any node could not connect to the master.
+- Fail if any node disconnected after a successful connection.
```
Comment (AdriiiPRodri) [inline tests/performance/test_cluster/test_cluster_logs/test_cluster_connection/README.md:98]:
```suggestion
- Fail if any node is disconnected after a successful connection.
```

File: tests/performance/test_cluster/test_cluster_logs/test_cluster_error_logs/test_cluster_error_logs.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+nodes_with_errors = {}
+
+
+# Functions
+def error_in_white_list(error_str):
+    for line in white_list:
+        if line in error_str:
+            return True
+
+    return False
```
Comment (AdriiiPRodri) [inline tests/performance/test_cluster/test_cluster_logs/test_cluster_error_logs/test_cluster_error_logs.py:25]:
```suggestion
    return True if any(line in error_str for line in white_list) else False
```

File: tests/performance/test_cluster/test_cluster_logs/test_cluster_error_logs/test_cluster_error_logs.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+from yaml import safe_load
+
+test_data_path = join(dirname(realpath(__file__)), 'data')
+white_list = [log.encode() for log in safe_load(open(join(test_data_path, 'configuration.yaml')))['white_list']]
+node_name = re.compile(r'.*/(master|worker_[\d]+)/logs/cluster.log')
+nodes_with_errors = {}
+
+
+# Functions
+def error_in_white_list(error_str):
```
Comment (AdriiiPRodri) [inline tests/performance/test_cluster/test_cluster_logs/test_cluster_error_logs/test_cluster_error_logs.py:20]: Missing docstring

File: tests/performance/test_cluster/test_cluster_logs/test_cluster_logs_order/README.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+### Tests information
+
+| Number of tests | Time spent |
+|:--:|:--:|
+| 1 | 0.2s |
+
+## Expected behavior
+
+- Fail if the any log is printed when it shouldn't.
```
Comment (AdriiiPRodri) [inline tests/performance/test_cluster/test_cluster_logs/test_cluster_logs_order/README.md:114]:
```suggestion
- Fail if any log is printed when it shouldn't.
```

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]

**Review by AdriiiPRodri — DISMISSED** [review]

**Review by davidjiglesias — APPROVED** [review]

File: tests/performance/test_cluster/test_cluster_logs/test_cluster_sync/README.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+## Objective
+
+To verify that:
+- Files are synced only once. 
+- Files are not deleted in the destination node.
+
+## General info
+### Parameters
+The test needs to receive one parameter in order to be run. If said parameter is not specified, the test will fail. The required parameter is:
```
Comment (Rebits) [inline tests/performance/test_cluster/test_cluster_logs/test_cluster_sync/README.md:22]:
```suggestion
The test needs to receive one parameter in order to be run. If this parameter is not specified, the test will fail. The required parameter is:
```

**Review by Rebits — CHANGES_REQUESTED** [review]
Awesome job, but some changes are required.

File: tests/performance/test_cluster/README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    │   └── logs
+    │       └── cluster.log
+    └── ...
+    ```
+- `--html=report.html`: Create a html report with the test results. 
+- `--self-contained-html`: Store all the necessary data for the report inside the html file.
+
+### Example output
+```shell
+python3 -m pytest . --artifacts_path='/home/selu/Descargas/cluster_performance/59' --n_agents=50000 --n_workers=10 --html=report.html --self-contained-html
```
[the remark this answers is not in the document]
Comment (Selutario) [reply tests/performance/test_cluster/README.md:49]: Done!

**Pull request merged**